### PR TITLE
Improve CBO estimates for correlated columns

### PR DESCRIPTION
**Pull request by raunaqmorarka**
## Description

Overall goal of the PR is to work towards enabling optimizer.default-filter-factor-enabled by default.
If default-filter-factor is enabled with existing implementation, it improves q18 and q21 on tpch significantly.
However, it also results in regressions on certain benchmark queries (tpcds partitioned q64, tpcds unpartitioned q78).
These changes update the estimation logic of filters and joins to address the problems
with underestimation of filter conjunctions and overestimation of multi-clause joins observed
when default-filter-factor is enabled with existing implementation.

> Is this change a fix, improvement, new feature, refactoring, or other?

Improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Query optimizer

> How would you describe this change to a non-technical end user or system administrator?

Improves CBO estimates in the presence of hard to estimate terms.

## Related issues, pull requests, and links

Picks first (n-1) commits from https://github.com/trinodb/trino/pull/11066

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Section
* Improve CBO estimates in the presence of correlated columns.
```

### CODE REVIEW

**Comment by raunaqmorarka** [comment]
TPC benchmark results with SF1000 ORC
Unpartitioned
[Correlation changes sf1000 orc unpartitioned.pdf](https://github.com/trinodb/trino/files/8185197/Correlation.changes.sf1000.orc.unpartitioned.pdf)
Partitioned
[Correlation changes sf1000 orc partitioned.pdf](https://github.com/trinodb/trino/files/8185199/Correlation.changes.sf1000.orc.partitioned.pdf)

**Review by sopel39 — APPROVED** [review]

**Comment by sopel39** [comment]
prefix already approved in https://github.com/trinodb/trino/pull/11066

**Comment by findepi** [comment]
> Improve CBO estimates for correlated columns

Is it about correlation in a sense of query outer scope references, or ... ?

**Comment by raunaqmorarka** [comment]
> > Improve CBO estimates for correlated columns
> 
> Is it about correlation in a sense of query outer scope references, or ... ?

This is about correlation in the data between columns (e.g. nation, city)

**Comment by raunaqmorarka** [comment]
Results with latest PR
[Correlated changes sf1000 orc partitioned.pdf](https://github.com/trinodb/trino/files/8222310/Correlated.changes.sf1000.orc.partitioned.pdf)
Ignore tpch/q05 result in partitioned run, the report is missing the number for that due to some glitch in benchmarking infra, there was no change to that query in the actual run.
[Correlation changes sf1000 unpartitioned orc.pdf](https://github.com/trinodb/trino/files/8222315/Correlation.changes.sf1000.unpartitioned.orc.pdf)

File: core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java [hunk 1]
```diff
@@ -442,4 +478,52 @@ private OptionalDouble doubleValueFromLiteral(Type type, Expression literal)
...
+
+    private static List<List<Expression>> extractCorrelatedGroups(List<Expression> terms, double filterConjunctionIndependenceFactor)
+    {
+        if (filterConjunctionIndependenceFactor == 1) {
+            // Allows the filters to be estimated as if there is no correlation between any of the terms
+            return ImmutableList.of(terms);
+        }
+
+        ArrayListMultimap<Expression, Symbol> expressionUniqueSymbols = ArrayListMultimap.create();
+        terms.forEach(expression -> expressionUniqueSymbols.putAll(expression, ImmutableList.copyOf(extractUnique(expression))));
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java:490]: I don't think you need to do `ImmutableList.copyOf`

File: core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java [hunk 2]
```diff
@@ -442,4 +478,52 @@ private OptionalDouble doubleValueFromLiteral(Type type, Expression literal)
...
+            symbolsPartitioner.find(expressionSymbols.get(0));
+            for (int i = 1; i < expressionSymbols.size(); i++) {
+                symbolsPartitioner.findAndUnion(expressionSymbols.get(0), expressionSymbols.get(i));
+            }
+        }
+
+        // Use disjoint sets of symbols to partition the given list of expressions
+        List<Set<Symbol>> symbolPartitions = ImmutableList.copyOf(symbolsPartitioner.getEquivalentClasses());
+        checkState(symbolPartitions.size() <= terms.size(), "symbolPartitions size exceeds number of expressions");
+        ArrayListMultimap<Integer, Expression> expressionPartitions = ArrayListMultimap.create();
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java:508]: `ArrayListMultimap<Integer, Expression>` -> `ListMultimap<Integer, Expression>`

**Review by sopel39 — APPROVED** [review]
small comments. Did tests changed?

File: core/trino-main/src/test/java/io/trino/cost/TestFilterStatsCalculator.java [hunk 3]
```diff
@@ -301,6 +311,118 @@ public void testAndStats()
...
+                        .distinctValuesCount(20.0)
+                        .nullsFraction(0.0))
+                .symbolStats("y", symbolAssert -> symbolAssert.averageRowSize(4.0)
+                        .lowValue(0.0)
+                        .highValue(5.0)
+                        .distinctValuesCount(20.0)
+                        .nullsFraction(0.1053779069));
+
+        assertExpression(
+                "x IN (0, 1, 2) AND (x = 0 OR (x = 1 AND y = 1) OR (x = 2 AND y = 1))",
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/cost/TestFilterStatsCalculator.java:413]: add a test for expression without any symbol

File: core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java [hunk 4]
```diff
@@ -442,4 +478,52 @@ private OptionalDouble doubleValueFromLiteral(Type type, Expression literal)
...
+        terms.forEach(expression -> expressionUniqueSymbols.putAll(expression, ImmutableList.copyOf(extractUnique(expression))));
+        // Partition symbols into disjoint sets such that the symbols belonging to different disjoint sets
+        // do not appear together in any expression.
+        DisjointSet<Symbol> symbolsPartitioner = new DisjointSet<>();
+        for (Expression term : terms) {
+            List<Symbol> expressionSymbols = expressionUniqueSymbols.get(term);
+            if (expressionSymbols.isEmpty()) {
+                continue;
+            }
+            symbolsPartitioner.find(expressionSymbols.get(0));
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java:499]: This one helps to add the symbol to DisjointSet when there is only one symbol in the list

[the remark this answers is not in the document]
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java:499]: Add a comment.

File: core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java [hunk 5]
```diff
@@ -442,4 +479,53 @@ private OptionalDouble doubleValueFromLiteral(Type type, Expression literal)
...
+            }
+            // Ensure that symbol is added to DisjointSet when there is only one symbol in the list
+            symbolsPartitioner.find(expressionSymbols.get(0));
+            for (int i = 1; i < expressionSymbols.size(); i++) {
+                symbolsPartitioner.findAndUnion(expressionSymbols.get(0), expressionSymbols.get(i));
+            }
+        }
+
+        // Use disjoint sets of symbols to partition the given list of expressions
+        List<Set<Symbol>> symbolPartitions = ImmutableList.copyOf(symbolsPartitioner.getEquivalentClasses());
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java:508]: I'm thinking we could avoid square root lookup here:
```
        Map<Symbol, Set<Symbol>> symbolPartitions = symbolsPartitioner.getEquivalentClasses().stream()
                .flatMap(set -> set.stream().map(symbol -> new SimpleEntry<>(symbol, set)))
                .collect(toImmutableMap(SimpleEntry::getKey, SimpleEntry::getValue));
        ListMultimap<Set<Symbol>, Expression> expressionPartitions = terms.stream()
                .filter(term -> !expressionUniqueSymbols.get(term).isEmpty())
                .collect(toImmutableListMultimap(term -> symbolPartitions.get(expressionUniqueSymbols.get(term).get(0)), term -> term));
        List<Expression> noSymbolExpressions = terms.stream()
                .filter(term -> !expressionUniqueSymbols.get(term).isEmpty())
                .collect(toImmutableList());
        return ImmutableList.<List<Expression>>builder()
                .addAll(expressionPartitions.keys().stream()
                        .map(expressionPartitions::get)
                        .iterator())
                .add(noSymbolExpressions)
                .build();
```

**Pull request merged**